agent/src/beerocks/slave/son_slave_thread.cpp

```diff
@@ -4944,46 +4944,51 @@ bool slave_thread::handle_channel_preference_query(Socket *sd, ieee1905_1::CmduM
 
     channel_preference_tlv->radio_uid() = network_utils::mac_from_string(config.radio_identifier);
 
-    // Create operating class object
-    auto op_class_channels = channel_preference_tlv->create_operating_classes_list();
-    if (!op_class_channels) {
-        LOG(ERROR) << "create_operating_classes_list() has failed!";
-        return false;
-    }
+    // Get supported channels and operating classes .
+    auto m = get_supported_channels_map();
 
     // TODO: check that the data is parsed properly after fixing the following bug:
     // Since sFlags is defined after dynamic list cPreferenceOperatingClasses it cause data override
     // on the first channel on the list and sFlags itself.
     // See: https://github.com/prplfoundation/prplMesh/issues/8
 
     // Fill operating class object
-    op_class_channels->operating_class() = 80; // random operating class for test purpose
+    for (auto it = m.begin(); it != m.end(); ++it) {
+        // Create operating class object
+        auto op_class_channels = channel_preference_tlv->create_operating_classes_list();
 
-    // Fill up channels list in operating class object, with test values
-    for (uint8_t ch = 36; ch < 50; ch += 2) {
-        // allocate 1 channel
-        if (!op_class_channels->alloc_channel_list()) {
-            LOG(ERROR) << "alloc_channel_list() has failed!";
+        if (!op_class_channels) {
+            LOG(ERROR) << "create_operating_classes_list() has failed!";
             return false;
         }
-        auto channel_idx = op_class_channels->channel_list_length();
-        auto channel     = op_class_channels->channel_list(channel_idx - 1);
-        if (!channel) {
-            LOG(ERROR) << "getting channel entry has failed!";
-            return false;
+
+        op_class_channels->operating_class() = it->first;
+        // Fill up channels list in operating class object
+        for (auto &ch : it->second) {
+            // allocate 1 channel
+            if (!op_class_channels->alloc_channel_list()) {
+                LOG(ERROR) << "alloc_channel_list() has failed!";
+                return false;
+            }
+            auto channel_idx = op_class_channels->channel_list_length();
+            auto channel     = op_class_channels->channel_list(channel_idx - 1);
+            if (!channel) {
+                LOG(ERROR) << "getting channel entry has failed!";
+                return false;
+            }
+            *channel = ch;
         }
-        *channel = ch;
-    }
 
-    // Update channel list flags
-    op_class_channels->flags().preference = 15; // channels preference
-    op_class_channels->flags().reason_code =
-        wfa_map::cPreferenceOperatingClasses::eReasonCode::UNSPECIFIED;
+        //TODO: get real preference score, and set Non-operable channel score to zero.
+        op_class_channels->flags().preference = 15; // channels preference
```

Comment:
The problem here is that you assume that the radio will always support all the channels on a specific operating class.
This is not will always be true.
For example, lets take the operating class 125 which contain channels {149, 153,157, 161,165, 169} with BW=20 MHz.
The radio can report it supports these channels, except for the last channel - 169.
In that case you need to add operating class 2 times:
1) With channel list {149, 153,157, 161,165} and preference "15".
2) With channel list {169} with preference "0" (non-operable).

Reply:
Maybe it's better to have list of operating classes with list of non-operable channels inside it. Radio Basic Caps should have list of non-operable channel in the operating class.